### PR TITLE
fix(agent): scope load_skill's advertised index to what the session can call

### PR DESCRIPTION
**Pull request by M3gA-Mind**
> **Stacked on #6214.** Base is `main` as usual, so this branch carries #6214's commit too — GitHub will show two. **Merge #6214 first**; this PR's diff then reduces to its own single commit (`f864f414c`). The overlap is real: both touch `toolpacks/{mod,tools,toolpacks_tests}.rs`, so this cannot apply to `main` on its own.

## Summary

- #6214 fixed the **landing**: `load_skill` on a pack you cannot use now returns the route instead of a schema dump. **The invitation was still wrong.**
- `pack_index_markdown()` is global and is baked into `LoadSkillTool`'s description at construction, before any session exists — so every agent read all ten packs as loadable, including ones it can call nothing in.
- Both halves of the advertisement are now scoped to the session: the description's index **and** the `skill` enum.
- The description gets **smaller**: 975 → 691 chars for the orchestrator, **~71 tokens saved every turn**, plus three fewer enum entries.

## Problem

For the orchestrator, three packs are advertised as loadable that it can call nothing in — `system` (26 tools), `documents` (2), `audio` (3). The model is told to go somewhere it cannot use, finds that out, and comes back.

Before #6214 that cost a dead turn. After #6214 it costs one wasted round trip. It should cost zero.

The carrier is the problem: `LoadSkillTool::new` builds the description once, and `skill_enum()` builds the enum once, both before any session exists.

## Solution

`visible_tool_specs_for_policy` (`session/builder/mod.rs:89`) is the only function that turns tools into the specs a provider sees **and** holds the session. It already clones every spec, so rewriting one is free, and it covers both the initial build and the Composio-reconcile rebuild.

#6214 moved pack *rendering* into the middleware for the same reason, but that seam is wrong here: a description ships with the schema **before** any tool call, so a call interceptor is too late.

Three details worth a reviewer's attention:

**The schema is the stronger half.** Narrowing the `skill` enum makes an unusable pack *unrepresentable*, not merely discouraged in prose. I would have wanted this even at zero token saving — and a shorter enum is fewer tokens, not more.

**It uses the gate's predicate, not the neighbour's.** `decision_for(..).is_denied()`, not the adjacent visibility filter's `is_allowed`. Matching the neighbour would re-introduce the two sources of truth #6214 exists to collapse.

**A session that can reach no pack loses both pack tools.** An empty index and an empty enum are not a tool, so `load_skill` and `use_skill` are dropped from the wire rather than shipped inert.

The `Unknown skill` error's listing is scoped by the same predicate — offering a hallucinating model a pack it cannot use is the same wrong turn, one error later.

## The design choice, and the tradeoff against it

An un-callable pack is **dropped** from the index rather than listed with an inline route (`` `system` — … · delegate to `manage_settings` ``). The inline form is the more obviously "honest" option, so here is why I did not take it:

**The capability is already advertised — by the delegate tool.** `manage_settings`, `do_crypto` and `make_presentation` are all on the wire with their `when_to_use` descriptions, and they are what the model should call. Listing the pack *and* its route duplicates routing the model already has, and bills it on **every turn**, to save a round trip the model should not be taking at all. The index answers "which skills can I load"; a pack that cannot be loaded is not an answer to that question.

**What this loses, stated against my own choice.** For a pack that is un-callable **and** has no reachable owner, the capability becomes invisible to that agent. That is exactly `audio` — ownerless (`owners: &[]`), so the orchestrator genuinely cannot do podcasts by any route, and advertising it only bought a wasted trip. I think invisibility is the correct representation of "cannot do this", but it is a real tradeoff, and `audio` is the one case where the inline-route option would have read better. A reviewer who disagrees should say so on that case specifically: swapping to inline routes is a change to `pack_index_markdown_filtered` alone.

Scoping is per-session, so `settings_agent` still sees `system`. Each agent gets its own truth.

**No pack was deleted, and I found none that should be.** "Dead" in the #6214 audit meant *dead to this caller*. `goals` and `app_update` are fully callable; `system` is doing exactly its job keeping 26 schemas out of the context window.

## What it cost

The advertisement shrank. Measured against the orchestrator's reachable set (`[tools].named` ∪ synthesised delegate names):

| | chars | packs |
|---|---:|---:|
| full index | 975 | 10 |
| scoped index | 691 | 7 |
| **saved** | **284 (~71 tokens/turn)** | dropped `system` (113c), `documents` (91c), `audio` (80c) |

## Verification

Four revert-checks, each failing on **my** assertion:

| revert | test | failure |
|---|---|---|
| description rewrite disabled | `the_index_drops_a_pack_this_session_can_call_nothing_in` | "`system` has no callable tool here and must not be advertised: …" |
| enum rewrite disabled | `the_skill_enum_offers_only_reachable_packs` | left: all ten pack ids; right: `["workflows"]` |
| never signal "drop it" | `a_session_that_can_reach_no_pack_loses_load_skill` | "with nothing reachable, load_skill must be dropped, not emptied" |
| **unwired from the builder** | `visible_specs_scope_load_skills_index_to_the_session` + `visible_specs_drop_the_pack_tools_when_no_pack_is_reachable` | "with no reachable pack, neither pack tool earns its schema: [load_skill, use_skill]" |

The fourth is the one that matters most and I added it for myself: it proves the helper is **wired**, not merely correct. A right helper nobody calls advertises all ten packs exactly as before, and the first three tests would still have passed.

`scoping_the_index_only_ever_shrinks_it` pins the token claim, so a future edit cannot quietly make the advertisement bigger.

**479 passed, 0 failed** across `toolpacks`, `builder_tests`, `orchestrator`, `tinyagents`, `session::builder`. `cargo fmt --check` clean; `cargo clippy -p openhuman --no-deps --all-targets` clean on every changed file.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement) — 7 new tests: index scoping, enum scoping, the no-reachable-pack drop, the shrink invariant, and two builder-level tests proving the wiring.
- [x] **Diff coverage ≥ 80%** — every changed production path is exercised and independently revert-checked: the description rewrite, the enum rewrite, the `false` return, and the `filter_map` in the builder. Full `pnpm test:coverage` not run — see Validation Blocked.
- [x] N/A: behaviour-only change — no feature row added, removed or renamed in `docs/TEST-COVERAGE-MATRIX.md`.
- [x] N/A: no coverage-matrix feature IDs apply (see the item above).
- [x] No new external network dependencies introduced (mock backend used per [Testing Strategy](../gitbooks/developing/testing-strategy.md#mock-policy)) — all new tests are in-process with a hand-built `ToolPolicySession`.
- [x] N/A: does not touch a release-cut surface in `docs/RELEASE-MANUAL-SMOKE.md` — this is agent-internal tool advertisement.
- [x] N/A: no issue exists — this follows from #6214's own audit finding, which was reported from a live turn rather than filed.

## Impact

- **Runtime/platform:** every agent that is offered `load_skill`. No RPC, schema or protocol change; no migration.
- **Behaviour change:** a pack with nothing callable in this session is no longer advertised, and is no longer a valid `skill` value. A session that can reach no pack at all carries neither `load_skill` nor `use_skill`.
- **Context cost falls** (~71 tokens/turn for the orchestrator); it can never rise, and a test pins that.
- **No behaviour change for an owner** — `settings_agent` still sees `system`, because it can call it.
- **Security:** strictly narrowing. The gate is untouched and still authoritative; the advertisement merely stopped disagreeing with it.

## Related

- Closes: N/A — no issue; follows from #6214's audit.
- Follow-up PR(s)/TODOs: `audio` is ownerless, so it is the one pack that now vanishes with no route offered anywhere — deliberately out of scope here and in #6214, because deciding who owns podcast generation is a product call. Likewise `composio`'s `integrations_agent` and `skills`' `context_scout`, which the orchestrator holds no delegate for.

---

### Linear Issue

- Key: N/A — not tracked in Linear.
- URL: N/A — follows from #6214's audit.

### Commit & Branch

- Branch: `fix/load-skill-index-scoping`
- Commit SHA: `f864f414cbb2af9fd0df87bc704b98aeb3d45f92` (stacked on `1939edeb6`, which is #6214)

### Validation Run

- [x] N/A: `pnpm --filter openhuman-app format:check` — no frontend files changed (Rust only).
- [x] N/A: `pnpm typecheck` — no TypeScript files changed.
- [x] Focused tests: `cargo test -p openhuman --lib --features "$(bash scripts/ci/product-features.sh)" --no-default-features -- toolpacks builder_tests orchestrator tinyagents session::builder` → **479 passed, 0 failed, 2 ignored**. Test-first + four revert-checks (table above).
- [x] Rust fmt/check (if changed): `cargo fmt --check` clean; `cargo clippy -p openhuman --no-deps --all-targets` clean on all changed files.
- [x] N/A: Tauri fmt/check — no files under the Tauri shell changed.

### Validation Blocked

- `command:` full-workspace `pnpm test:rust` / `pnpm test:coverage`
- `error:` not run — the fleet is capped at 3 concurrent local cargo slots. `RUST_MIN_STACK=67108864` is required near the agent harness or an unrelated turn test overflows its stack.
- `impact:` diff coverage is argued from the revert-checks rather than a `diff-cover` number. CI's gate is authoritative.

### Behavior Changes

- Intended behavior change: `load_skill`'s advertised index and `skill` enum are scoped to the packs the calling session can actually use.
- User-visible effect: the assistant stops making a round trip into a skill it cannot use, and each turn carries a slightly smaller tool schema.

### Parity Contract

- Legacy behavior preserved: `pack_index_markdown()` still exists and still returns the full index for callers with no session; an owner's advertisement is unchanged because nothing it can call is removed.
- Guard/fallback/dispatch parity checks: the advertisement now uses the same predicate as the gate and as #6214's listing filter — one source of truth for "can this session call it" across all three. Degradation is one-directional: a pack is dropped only when *every* tool in it is denied.

### Duplicate / Superseded PR Handling

- Duplicate PR(s): N/A.
- Canonical PR: this one; #6214 is its prerequisite, not a duplicate.
- Resolution: N/A.

## Summary by CodeRabbit

- **New Features**
  - Skill listings now show only tools available to the current session.
  - Unavailable skills are omitted from advertised options, with guidance for reaching an appropriate available route.
  - Permission checks now apply consistently when loading or using skills.
  - Workflow requests are routed to dedicated building and discovery tools, while running and listing saved workflows remains available.

- **Bug Fixes**
  - Prevented users from being shown tools they cannot call.
  - Improved fallback guidance when no directly usable workflow tools are available.

### CODE REVIEW

**Comment by M3gA-Mind** [comment]
⚠️ Two backstop findings — W2 still owns both, I have not touched the branch.

**1. This PR contains #6214, so it inherits #6214's compile break.** `605fc7fa6` is an ancestor of this head (`2dd4dbd6d`), so this is a true stack, not two branches off `main`:

```
$ git log --oneline upstream/main..2dd4dbd6d
2dd4dbd6d fix(agent): scope load_skill's advertised index to what the session can call
605fc7fa6 fix(agent): make load_skill's listing agree with the gate that refuses it
```

`main + 2dd4dbd6d` fails `cargo check … --test raw_coverage_all` with the same `DelegationTarget: From<&str>` error I reported on #6214, and `main` alone passes that check. Fixing it on #6214 fixes it here.

**2. The stacking has consequences worth deciding deliberately, because both PRs say `base: main`:**

- Reviewers of this PR are reading **both** changes — the diff shown includes #6214's.
- **Merge order matters.** If #6214 merges first, this one follows normally. If this one merges first, it takes #6214's commit with it and #6214 becomes an empty no-op that still looks open.
- The "clean merge of each" reassurance does not apply between these two — they are not independent.

Either retarget this PR's base to `fix/orchestrator-workflow-routing` so the stack is visible to GitHub and the diff narrows to just this change, or land #6214 first and leave a note here. Your call; I am flagging it rather than picking.

**Not a collision with #6217, which I did check.** #6215 and #6217 both touch `toolpacks/registry.rs` (different regions: `packed_tool_names_for_agent` vs the `PACKS` const) and all three touch `toolpacks_tests.rs`. Merging all three onto `main` is textually clean, and the only compile failure in that merged tree is the one above — inherited, not an interaction. So there is no #6217 interference to resolve, only the #6214 break.

File: src/openhuman/agent/harness/session/builder/mod.rs [hunk 1]
```diff
@@ -91,13 +91,44 @@ pub(super) fn visible_tool_specs_for_policy(
...
+    // any session exists, so every agent was told all ten packs were loadable —
+    // including ones it can call nothing in. The model went, found out, and came
+    // back.
+    //
+    // This is the only place that turns tools into the specs a provider sees AND
+    // holds the session, so it is where the invitation is made to agree with the
+    // gate. The same predicate the gate uses (`decision_for(..).is_denied()`),
+    // not `is_allowed` — matching the filter above would re-introduce the two
+    // sources of truth this whole fix exists to collapse.
+    let is_callable = |name: &str| !tool_policy.decision_for(name).is_denied();
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/openhuman/agent/harness/session/builder/mod.rs]: Correct, and this was the more dangerous of the two — thank you.

You're right that `strip_packed_from_visible` runs before the policy snapshot is built, so every non-owner packed tool is `HideFromPrompt`, and `is_denied()` treats that as denied. My predicate therefore reported *no* pack as reachable, which meant `scope_load_skill_spec` returned `false` and my own `filter_map` dropped **both** `load_skill` and `use_skill` from the wire — deleting the only route to every withheld tool, which is the precise opposite of what this PR is for.

My tests did not catch it because they hand-built a `ToolPolicySession` whose `decisions` were explicit `Allow`s. That fixture cannot express the state a packed tool is actually in, so it agreed with itself. The new test builds the session the way the harness does:

```rust
let mut visible: HashSet<String> = tools.iter().map(|t| t.name().to_string()).collect();
strip_packed_from_visible(&mut visible, "orchestrator");
let session = ToolPolicyEngine::build_session("orchestrator", "web_chat", "chat",
                                              &Default::default(), &tools, &visible);
```

and asserts `load_skill`, `use_skill` and the `goals` pack all survive.

**Fix.** `ToolPolicyDecision::blocks_execution()` (added on #6214, the prerequisite) separates the disclosure mask from an execution denial: `Deny | RequireApproval` block; `HideFromPrompt` does not. The index predicate here now uses it.

On your suggested shape — "determine callability from the channel/permission policy before applying the disclosure mask". I went with distinguishing the two states rather than reordering, for two reasons. The information is already in the snapshot, so no reordering is needed to recover it; and the ordering is load-bearing elsewhere (the visibility filter immediately above this line genuinely does want the post-mask answer). Naming the distinction fixes it once for all four call sites, including the pre-existing `use_skill` gate bug on #6214, which a reordering local to this function would not have touched. If you think the reorder is still worth it on top, say so and I will look again.

Revert-check — restoring `blocks_execution` to `is_denied`:

```
a_realistic_withheld_session_keeps_its_packs_advertised ... FAILED
    load_skill must survive — the pack it opens is reachable: []
```

**493 passed, 0 failed.** Head is now `451a178e7`, rebased on #6214's new head `dbdad815c`.

**Comment by M3gA-Mind** [comment]
## Force-pushed `2dd4dbd6d` → `451a178e7`

Rebased on #6214's new head `dbdad815c`, and fixes Codex's P1 ([thread](https://github.com/tinyhumansai/openhuman/pull/6215#discussion_r3987072280)).

**The bug he caught was severe and mine.** Callability was computed with `is_denied()`, which is true for `HideFromPrompt` — the state every withheld packed tool is in. So no pack read as reachable, `scope_load_skill_spec` returned `false`, and my own `filter_map` dropped **both `load_skill` and `use_skill` from the wire** — deleting the only route to every withheld tool. The exact opposite of this PR's intent.

**Why my tests missed it:** they hand-built a `ToolPolicySession` with explicit `Allow` decisions, a fixture that cannot express the real classification, so it agreed with itself.

**What changed**

- The index predicate now uses `ToolPolicyDecision::blocks_execution()` (added on #6214).
- New `a_realistic_withheld_session_keeps_its_packs_advertised` builds the session through `ToolPolicyEngine::build_session` after `strip_packed_from_visible`, and asserts `load_skill`, `use_skill` and the `goals` pack all survive.
- `toolpacks_tests.rs` split at the 750-line layout limit.

**Revert-check:** restoring `is_denied` fails it with `load_skill must survive — the pack it opens is reachable: []` — the empty list is exactly the predicted outcome.

**493 passed, 0 failed.** fmt, clippy, layout clean. No test weakened or deleted.

File: src/openhuman/agent/tinyagents/middleware_part_02.rs [hunk 2]
```diff
@@ -561,6 +561,83 @@ impl ToolPolicyMiddleware {
...
+            let Some(target) =
+                crate::openhuman::tools::traits::delegation_target(tool.as_ref())
+            else {
+                continue;
+            };
+            if !owners.contains(&target) {
+                continue;
+            }
+            let name = tool.name().to_string();
+            if self.session.decision_for(&name).blocks_execution() || found.contains(&name) {
```
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/openhuman/agent/tinyagents/middleware_part_02.rs]: Correct, fixed as proposed, with the regression test you asked for.

You caught the genuine interaction. Once `blocks_execution()` returns `false` for `HideFromPrompt`, `callable_delegates_for` starts including prompt-hidden delegates, and `route_sentence` advertises a **direct** call — while the direct-call gate still rejects `HideFromPrompt` via `is_denied()`. The model would be steered into a second refusal one hop after the first.

It is worth naming *why* the two predicates differ rather than just switching one, because the asymmetry looks like an inconsistency and is the opposite:

| path | gate predicate | hint must use |
|---|---|---|
| `use_skill { tool: X }` | `blocks_execution()` — hidden is fine, that route *is* the disclosure mechanism | `blocks_execution()` |
| direct call to `X` | `is_denied()` — a tool the model was never shown must not be callable by name | `is_denied()` |

**A hint has to use the predicate of the gate it points at.** `route_sentence` points at a direct call, so it takes `is_denied()`. I left a comment at the call site saying exactly that, since the next reader will otherwise "fix" the inconsistency back.

I did not reach for the other resolution — relaxing the direct-call gate to accept hidden tools — and I want that on the record. It would make the hint true by widening the gate, defeating the disclosure guarantee (`host/security_gate.rs:239`: *"a tool the model was never shown"*). Narrowing the hint is the correct direction.

Regression test `a_prompt_hidden_delegate_is_not_offered_as_a_direct_route`: builds a real session (`strip_packed_from_visible` → `ToolPolicyEngine::build_session`) in which `build_workflow` is itself a `workflows` pack member and therefore prompt-hidden, then asserts the route falls back to the owning agent.

Revert-check — putting `blocks_execution()` back:

```
a_prompt_hidden_delegate_is_not_offered_as_a_direct_route ... FAILED
    a prompt-hidden delegate must not be advertised as a direct call:
    Call `build_workflow` instead — that agent owns these tools and runs them directly.
```

The fix is in `middleware_part_02.rs`, which lives in #6214 (the prerequisite), so it landed there — head `1207db773`. This PR is rebased on it at `ff7906589`. **495 passed, 0 failed.**

File: src/openhuman/agent/harness/session/builder/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (M3gA-Mind) [reply src/openhuman/agent/harness/session/builder/mod.rs]: Follow-up now that the fix has landed. **Your finding and the one on #6214 are the same bug at two sites**, and I want that explicit rather than resolving them separately as if they were unrelated.

One root cause: `HideFromPrompt` — the normal state of every withheld packed tool — was being read as a denial, because `is_denied()` is `!matches!(action, Allow)`.

| site | PR | symptom |
|---|---|---|
| `channel_permission_block` inner `use_skill` check | #6214 | refused every withheld packed tool — **pre-existing on `main`**, not a branch regression |
| `render_pack_filtered` listing predicate | #6214 | listing rendered empty, "no tools available" for every pack |
| `scope_load_skill_spec` index predicate | #6215 (here) | no pack reachable → `load_skill` *and* `use_skill` dropped from the wire |

One fix covers all three: `ToolPolicyDecision::blocks_execution()` (`agent_policy/types.rs`, added on #6214, which this PR is rebased on). `Deny | RequireApproval` block execution; `HideFromPrompt` does not; and it carries the permission ceiling, because `build_session_from_refs` tests `explicitly_hidden` before `exceeds_permission` and so never records a verdict for a tool that is both (a separate CodeRabbit Critical).

The site here is at `session/builder/mod.rs`, head `ff7906589`, pinned by `a_realistic_withheld_session_keeps_its_packs_advertised` — which builds the session the way the harness does rather than the hand-written `Allow` fixture that let my original version pass.

**495 passed, 0 failed.**

**Comment by M3gA-Mind** [comment]
## Force-pushed `451a178e7` → `ff7906589`

Rebased on #6214's new head `1207db773`, which carries two accepted review fixes: the permission ceiling in `blocks_execution()` (CodeRabbit Critical) and the route hint switching to `is_denied()` (CodeRabbit Minor, filed here — the code lives in `middleware_part_02.rs`, so the fix landed on the prerequisite).

No change to this PR's own diff. Both threads here are answered and resolved; the Codex P1 reply spells out that it and the one on #6214 are **one root cause at three sites**, so a reviewer can see them as a single bug rather than two coincidences.

**495 passed, 0 failed.** fmt, clippy, layout gate clean.

**Comment by M3gA-Mind** [comment]
## Rebased `ff7906589` → `196b33323`

No change to this PR's own diff — rebased onto #6214's new head `b43642a84`, which carries two consistency fixes: a rewritten predicate comment (tinysweeper) and a de-vacuumed test assertion (CodeRabbit). Details on #6214.

Both threads here remain answered and resolved. **495 passed, 0 failed.**

**Comment by M3gA-Mind** [comment]
## Rebased onto `upstream/main` — `f7e95b6e8` → `fbe9e0a2f`

Was ~23 commits behind; its lanes had last run against a base that no longer exists. Now **1 ahead, 0 behind** `10d2f1c67`.

**No code change** — clean rebase, no conflicts, single commit, signed, no attribution trailers. The diff against the new base is the same change it always was: `load_skill`'s advertised index and `skill` enum scoped to the packs the calling session can actually use.

**Now standalone.** #6214 merged, so this branch no longer carries that commit — the prerequisite is in `main` rather than stacked here.

Re-verified on the new base after re-initialising submodules: **442 passed, 0 failed** across `toolpacks`, `builder_tests`, `session::builder`, `tinyagents`, `agent_policy`; layout gate passes.

### One inherited failure to expect, which is not this PR's

`upstream/main` currently does not compile `--test raw_coverage_all`:

```
error[E0277]: the trait bound `DelegationTarget: From<&str>` is not satisfied
  --> tests/raw_coverage/inference_agent_raw_coverage_e2e.rs:3086:32
```

That came in with #6214 and is fixed by **#6222**, which is open. Anything based on main inherits it until that merges. If a lane here goes red on that error it is inherited, not introduced by this PR.
